src/librustc_mir/util/liveness.rs

```diff
@@ -0,0 +1,225 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Liveness analysis which computes liveness of MIR local variables at the boundary of basic blocks
+
+use rustc::mir::*;
+use rustc::mir::visit::{LvalueContext, Visitor};
+use rustc_data_structures::indexed_vec::{IndexVec, Idx};
+use rustc_data_structures::indexed_set::IdxSetBuf;
+use util::pretty::{write_basic_block, dump_enabled, write_mir_intro};
+use rustc::mir::transform::MirSource;
+use rustc::ty::item_path;
+use std::path::{PathBuf, Path};
+use std::fs;
+use rustc::ty::TyCtxt;
+use std::io::{self, Write};
+
+pub type LocalSet = IdxSetBuf<Local>;
+
+#[derive(Eq, PartialEq, Clone)]
+struct BlockInfo {
+    defs: LocalSet,
+    uses: LocalSet,
+}
+
+struct BlockInfoVisitor {
+    pre_defs: LocalSet,
+    defs: LocalSet,
+    uses: LocalSet,
+}
+
+impl<'tcx> Visitor<'tcx> for BlockInfoVisitor {
+    fn visit_lvalue(&mut self,
+                    lvalue: &Lvalue<'tcx>,
+                    context: LvalueContext<'tcx>,
+                    location: Location) {
+        if let Lvalue::Local(local) = *lvalue {
+            match context {
+                LvalueContext::Store |
+
+                // We let Call defined the result in both the success and unwind cases.
+                // This may not be right.
+                LvalueContext::Call |
+
+                // Storage live and storage dead aren't proper defines, but we can ignore
+                // values that come before them.
+                LvalueContext::StorageLive |
+                LvalueContext::StorageDead => {
+                    self.defs.add(&local);
+                }
+                LvalueContext::Projection(..) |
+
+                // Borrows only consider their local used at the point of the borrow.
+                // This won't affect the results since we use this analysis for generators
+                // and we only care about the result at suspension points. Borrows cannot
+                // cross suspension points so this behavoir is unproblematic.
```

Comment:
Maybe worth moving this comment to the top-level? Say?
```Rust
//! This analysis considers references as being used only at the point of the borrow. This
//! means that this does not track uses because of references that already exist:
//!
//! ```Rust
//!     fn foo() {
//!         x = 0;
//!         // `x` is live here
//!         GLOBAL = &x: *const u32;
//!         // but not here, even while it can be accessed through `GLOBAL`.
//!         foo();
//!         x = 1;
//!         // `x` is live again here, because it is assigned to `OTHER_GLOBAL`
//!         OTHER_GLOBAL = &x: *const u32;
//! }
//!
//! This means that users of this analysis still have to check whether pre-existing
//! references can be used to access the value (e.g. at movable generator yield points,
//! all pre-existing references are invalidated, so this doesn't matter).
```